### PR TITLE
Fix propagate partition properties for full outer join

### PR DESCRIPTION
**Pull request by kokosing**
Fix propagate partition properties for full outer join

### CODE REVIEW

**Comment by kokosing** [comment]
Extracted from https://github.com/prestosql/presto/pull/262

**Comment by martint** [comment]
This needs a better commit message. It’s not clear from the commit what’s broken, what kinds of queries it affects, etc.

**Comment by kokosing** [comment]
Improved a commit message a bit.

**Comment by sopel39** [comment]
I have little context to the change. I think @martint knows the most here

File: presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PropertyDerivations.java [hunk 1]
```diff
@@ -866,7 +866,7 @@ else if (equality.getRight().equals(column) && columns.contains(equality.getLeft
                         .filter(SymbolReference.class::isInstance)
                         .map(Symbol::from)
                         .collect(toImmutableSet());
-                if (symbolsInAssignment.containsAll(symbolInExpression)) {
+                if (symbolsInAssignment.equals(symbolInExpression)) {
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PropertyDerivations.java]: This is not enough. It needs to make sure the arguments to both coalesce expressions are *only* simple symbol references. Here are a couple of queries that don't produce the right results:

```sql
SELECT coalesce(t.a, s.a + 1 - 1, r.a)
FROM (VALUES 1, 2, 3) t(a)
FULL OUTER JOIN (VALUES 1, 4) s(a) ON t.a = s.a
FULL OUTER JOIN (VALUES 2, 5) r(a) ON t.a = r.a
GROUP BY 1
```

```sql
SELECT coalesce(t.a, u.a, if(t.b is null, 100, t.b)), count(*)
FROM (VALUES (1, 10), (2, 20), (3, 30), (null, 40), (100, 50)) t(a, b)
FULL OUTER JOIN (VALUES 1, 4, null) u(a) ON t.a = u.a
GROUP BY 1;
```

File: presto-main/src/test/java/io/prestosql/sql/planner/optimizations/TestFullOuterJoinWithCoalesce.java [hunk 2]
```diff
@@ -124,4 +125,27 @@ public void testFullOuterJoinWithCoalesceNonSymbol()
                                                                         exchange(LOCAL, GATHER, anyTree(values(ImmutableList.of("s")))))),
                                                         exchange(LOCAL, GATHER, anyTree(values(ImmutableList.of("r")))))))))));
     }
+
+    @Test
+    public void testFullOuterJoinWithCoalesceOnIfExpressionAndCountAggregation()
```
Comment (kokosing) [inline presto-main/src/test/java/io/prestosql/sql/planner/optimizations/TestFullOuterJoinWithCoalesce.java]: @martint Please verify if the expected plan is correct.

**Comment by kokosing** [comment]
@martint Comments addressed (thanks to @rongrong).

**Comment by martint** [comment]
Superseded by https://github.com/prestosql/presto/pull/288

**Pull request closed without merging**